### PR TITLE
refactor: Reactivate workflow locking

### PR DESCRIPTION
**Pull request by krynble**

### CODE REVIEW

File: packages/cli/src/workflows/workflows.services.ts [hunk 1]
```diff
@@ -206,6 +207,17 @@ export class WorkflowsService {
 			);
 		}
 
+		// Update the workflow's hash
+		workflow.hash = uuid();
```
Comment (netroy) [inline packages/cli/src/workflows/workflows.services.ts:211]: If we use `@BeforeUpdate()` to set this in `WorkflowEntity`, then we ca be certain that this get updated on every update, irrespective of where it comes from in the code.

Comment (krynble) [reply packages/cli/src/workflows/workflows.services.ts:211]: Fair point, but we have another PR to be merged (#4200 ) and it will break if we do so. We'll be updating the `workflow_entity` on other stuff than only workflow editions, leading to more false positives. For this reason I think it's best to only update the hash when actual changes happen.

File: packages/editor-ui/src/views/NodeView.vue [hunk 2]
```diff
@@ -3367,6 +3367,7 @@ export default mixins(
 			dataPinningEventBus.$off('pin-data', this.addPinDataConnections);
 			dataPinningEventBus.$off('unpin-data', this.removePinDataConnections);
 			nodeViewEventBus.$off('saveWorkflow', this.saveCurrentWorkflowExternal);
+			this.workflowsStore.setWorkflowId(PLACEHOLDER_EMPTY_WORKFLOW_ID);
```
Comment (ivov) [inline packages/editor-ui/src/views/NodeView.vue:3370]: Why was this needed? Here and at L3379.

File: packages/cli/src/databases/migrations/mysqldb/1669739707125-AddWorkflowHashColumn.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+import { logMigrationEnd, logMigrationStart } from '@db/utils/migrationHelpers';
+import config from '@/config';
+import { v4 as uuidv4 } from 'uuid';
+
+export class AddWorkflowHashColumn1669739707125 implements MigrationInterface {
+	name = 'AddWorkflowHashColumn1669739707125';
+
+	async up(queryRunner: QueryRunner): Promise<void> {
```
Comment (ivov) [inline packages/cli/src/databases/migrations/mysqldb/1669739707125-AddWorkflowHashColumn.ts:9]: Inferrable:
```suggestion
	async up(queryRunner: QueryRunner) {
```

File: packages/cli/src/databases/migrations/mysqldb/1669739707125-AddWorkflowHashColumn.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+			`ALTER TABLE ${tablePrefix}workflow_entity ADD COLUMN hash CHAR(36)`,
+		);
+
+		const workflowIds: Array<{ id: number }> = await queryRunner.query(`
+			SELECT id
+			FROM ${tablePrefix}workflow_entity
+		`);
+
+		workflowIds.map(({ id }) => {
+			const [updateQuery, updateParams] = queryRunner.connection.driver.escapeQueryWithParameters(
```
Comment (ivov) [inline packages/cli/src/databases/migrations/mysqldb/1669739707125-AddWorkflowHashColumn.ts:24]: Some UUIDs might need escaping? Or why do we need this?

File: packages/cli/src/databases/entities/WorkflowEntity.ts [hunk 5]
```diff
@@ -90,32 +85,8 @@ export class WorkflowEntity extends AbstractEntity implements IWorkflowDb {
 	})
 	pinData: ISimplifiedPinData;
 
-	/**
-	 * Hash of editable workflow state.
-	 */
+	@Column({ length: 36 })
```
Comment (ivov) [inline packages/cli/src/databases/entities/WorkflowEntity.ts:88]: We are no longer hashing, so `versionId` instead of `hash`?

File: packages/cli/src/workflows/workflows.controller.ee.ts [hunk 6]
```diff
@@ -213,7 +216,7 @@ EEWorkflowController.patch(
 	'/:id(\\d+)',
 	ResponseHelper.send(async (req: WorkflowRequest.Update) => {
 		const { id: workflowId } = req.params;
-		// const forceSave = req.query.forceSave === 'true'; // disabled temporarily - tests were also disabled
+		const forceSave = req.query.forceSave === 'true';
```
Comment (ivov) [inline packages/cli/src/workflows/workflows.controller.ee.ts:219]: If you can please unskip and update the relevant tests, and ensure they pass on all three DBs.

**Review by ivov — COMMENTED** [review]
No issues so far on manual testing.

Two UX questions:

1. Blocker message is way too long - display duration does not allow to read it in full at normal pace, even when expecting it. And even if we extend duration, blocker message is still probably too long for a simple toast message. Also in the blocker message it would be nice to include a link to open the workflow in a separate tab.
2. I noticed at `/workflows`, if a user activates a workflow, a sharee is still able to freely activate/deactivate that same workflow at `/workflows`, and vice versa. Is this expected? Should they not be blocked?

File: packages/cli/src/databases/migrations/mysqldb/1669739707125-AddWorkflowHashColumn.ts (hunk 4 above)
Comment (krynble) [reply packages/cli/src/databases/migrations/mysqldb/1669739707125-AddWorkflowHashColumn.ts]: No special reason, it's just good practice to send queries through the escape mechanism

Comment (netroy) [reply packages/cli/src/databases/migrations/mysqldb/1669739707125-AddWorkflowHashColumn.ts]: doesn't the query-runner do that automatically in the next line?

File: packages/editor-ui/src/views/NodeView.vue (hunk 2 above)
Comment (cstuncsik) [reply packages/editor-ui/src/views/NodeView.vue:3370]: Otherwise last active workflow ID stuck in the store. In my thinking, if the user is not on the workflow page there should not be active workflow ID, right? 

This stuck ID caused some trouble on the workflow list page.

Comment (ivov) [reply packages/editor-ui/src/views/NodeView.vue:3370]: Thank you, makes sense.

File: packages/cli/src/workflows/workflows.controller.ee.ts (hunk 6 above)
Comment (krynble) [reply packages/cli/src/workflows/workflows.controller.ee.ts:219]: Thanks for reminding me this, tests uncommented and fixed.

File: packages/cli/src/databases/entities/WorkflowEntity.ts (hunk 5 above)
Comment (krynble) [reply packages/cli/src/databases/entities/WorkflowEntity.ts:88]: Good idea, changed!

**Review by ivov — APPROVED** [review]
Thank you for fixing this. Tested manually and on all three DBs. Only two minor comments that should not block merging so approving.

**Comment by janober** [comment]
Got released with `n8n@0.206.1`
